File: fsharp-backend/src/LibBackend/password.ml
type t = string

let to_bytes pw = pw

let from_plaintext password =
  password
  |> Bytes.of_string
  |> Libcrypto.Hash.wipe_to_password
  |> Libcrypto.Hash.hash_password Sodium.Password_hash.interactive
  |> Bytes.to_string
  |> B64.encode


let invalid = ""

let from_hash pw = pw

File: fsharp-backend/src/LibBackend/Password.fs
module LibBackend.Password

open Prelude

// FSTODO: this does not yet match the OCaml version. It's not strictly
// necessary that this version does match, as it's not really used. However,
// Crypto::password function use the same algorithm so we do need to get this
// right.

type T =
  private
  | Pw of string

  override this.ToString() : string =
    let (Pw pw) = this
    pw

let fromPlaintext (password : string) : T =
  password
  |> Sodium.PasswordHash.ArgonHashString
  |> UTF8.toBytes
  |> Base64.defaultEncodeToString
  |> Pw

let invalid : T = Pw ""

let fromHash (hash : string) : T = Pw hash
